x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -177,42 +177,15 @@ protected static List<Batch<LogicalPlan>> rules() {
         return asList(substitutions, operators(), skip, cleanup(), defaultTopN, label);
     }
 
-    static class InvertSpatialFunctions extends OptimizerRules.OptimizerRule<UnaryPlan> {
+    static class SubstituteSpatialSurrogates extends OptimizerRules.OptimizerExpressionRule<SpatialRelatesFunction> {
 
-        InvertSpatialFunctions() {
+        SubstituteSpatialSurrogates() {
             super(TransformDirection.UP);
         }
 
         @Override
-        protected LogicalPlan rule(UnaryPlan plan) {
-            if (plan instanceof Filter filter) {
-                var condition = filter.condition().transformDown(SpatialRelatesFunction.class, this::invertIfNecessary);
-                if (filter.condition().equals(condition) == false) {
-                    plan = new Filter(filter.source(), filter.child(), condition);
-                }
-            }
-            if (plan instanceof Eval eval) {
-                List<Alias> fields = eval.fields();
-                List<Alias> changed = fields.stream()
-                    .map(f -> (Alias) f.transformDown(SpatialRelatesFunction.class, this::invertIfNecessary))
-                    .toList();
-                if (changed.equals(fields) == false) {
-                    plan = new Eval(eval.source(), eval.child(), changed);
-                }
-            }
-            return plan;
-        }
-
-        private SpatialRelatesFunction invertIfNecessary(SpatialRelatesFunction srf) {
-            if (srf.isCommutative()) {
-                // Never bother to invert commutative functions
-                return srf;
-            }
-            if (srf.left().foldable() && srf.right().foldable() == false) {
-                // If we have a constant on the left, and a field on the right, invert the function
-                return srf.invert();
-            }
-            return srf;
+        protected SpatialRelatesFunction rule(SpatialRelatesFunction function) {
+            return function.surrogate();
```

Comment:
Looks good in the new version!

Reply:
Thanks, and I am adding a comment explaining the similarity to the two other approaches, and the decision to go with this approach for now.